server/src/test/java/org/apache/druid/server/QuerySchedulerTest.java

```diff
@@ -0,0 +1,566 @@
+            Assert.assertEquals(2, scheduler.getLaneAvailableCapacity(HiLoQueryLaningStrategy.LOW));
+          }
+        });
+        Sequence<Integer> results = scheduler.run(scheduled, underlyingSequence);
+        int rowCount = consumeAndCloseSequence(results);
+
+        Assert.assertEquals(10, rowCount);
+      }
+      catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
+    });
+    future.get();
+    Assert.assertEquals(5, scheduler.getTotalAvailableCapacity());
+  }
+
+  @Test
+  public void testHiLoLo() throws ExecutionException, InterruptedException
+  {
+    TopNQuery report = makeReportQuery();
+    ListenableFuture<?> future = executorService.submit(() -> {
+      try {
+        Query<?> scheduledReport = scheduler.laneQuery(QueryPlus.wrap(report), ImmutableSet.of());
+        Assert.assertNotNull(scheduledReport);
+        Assert.assertEquals(HiLoQueryLaningStrategy.LOW, QueryContexts.getLane(scheduledReport));
+
+        Sequence<Integer> underlyingSequence = makeSequence(10);
+        underlyingSequence = Sequences.wrap(underlyingSequence, new SequenceWrapper()
+        {
+          @Override
+          public void before()
+          {
+            Assert.assertEquals(4, scheduler.getTotalAvailableCapacity());
+            Assert.assertEquals(1, scheduler.getLaneAvailableCapacity(HiLoQueryLaningStrategy.LOW));
+          }
+        });
+        Sequence<Integer> results = scheduler.run(scheduledReport, underlyingSequence);
+
+        int rowCount = consumeAndCloseSequence(results);
+        Assert.assertEquals(10, rowCount);
+      }
+      catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
+    });
+    future.get();
+    Assert.assertEquals(5, scheduler.getTotalAvailableCapacity());
+    Assert.assertEquals(2, scheduler.getLaneAvailableCapacity(HiLoQueryLaningStrategy.LOW));
+  }
+
+  @Test
+  public void testHiLoReleaseSemaphoreWhenSequenceExplodes() throws Exception
+  {
+    expected.expectMessage("exploded");
+    expected.expect(ExecutionException.class);
+    TopNQuery interactive = makeInteractiveQuery();
+    ListenableFuture<?> future = executorService.submit(() -> {
+      try {
+        Query<?> scheduled = scheduler.laneQuery(QueryPlus.wrap(interactive), ImmutableSet.of());
+
+        Assert.assertNotNull(scheduled);
+
+        Sequence<Integer> underlyingSequence = makeExplodingSequence(10);
+        underlyingSequence = Sequences.wrap(underlyingSequence, new SequenceWrapper()
+        {
+          @Override
+          public void before()
+          {
+            Assert.assertEquals(4, scheduler.getTotalAvailableCapacity());
+          }
+        });
+        Sequence<Integer> results = scheduler.run(scheduled, underlyingSequence);
+
+        consumeAndCloseSequence(results);
+      }
+      catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
+    });
+    future.get();
+    Assert.assertEquals(5, scheduler.getTotalAvailableCapacity());
```

Comment:
This assert is not executed. Maybe use an exception handler instead of `expected`?

Reply:
Ah, i think that was a leftover from copying another test to make this one, but I think it does seem useful to assert that the token was released after the exception so will change